# X3FL dynamic relationships

## The relationship

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns

In [2]:
all_races = ["Teladi", "NMMC", "Goner", "TerraCorp", "Strong Arms", "Argon", "Boron", "Split", "Atreus", "OTAS", "Duke's", "Paranid", "Pirates", "Terran", "Yaki"]
# https://imgur.com/nqI0nbO, https://imgur.com/7BhbUMa, by blazenclaw, forum.egosoft.com
try:
    relationship_df = pd.read_csv("relationship.csv", index_col=0)
except:
    relationship_df = pd.read_csv("https://raw.githubusercontent.com/mkmark/X3FL-dynamic-relationships/main/relationship.csv", index_col=0)
relationship_df = pd.DataFrame(relationship_df)
cmap = sns.diverging_palette(5, 250, as_cmap=True)

def color_map(val):
    if val > 0.3:
        val = 0.3
    if val< -0.3:
        val = -0.3
    val = val/0.45+0.5
    cmap_val = cmap(val)
    colorvalue = (int(255*cmap_val[0]),int(255*cmap_val[1]),int(255*cmap_val[2]),cmap_val[3])
    color = 'rgba%s' % str(colorvalue)
    return 'background-color: %s' % color

relationship_df.style.applymap(color_map, subset=pd.IndexSlice[all_races, all_races])\
    .set_properties(**{'width': '40px', 'height': '40px'})\
    .set_caption("Relationship")\
    .set_precision(2)

The problem is simple: given the relationship $\mathbf{R}$, target notority $\mathbf{N}$, find tactics $\mathbf{X}$, which is a vector that represents how much notority point one should get from a race.

$$
\mathbf{R}\mathbf{X}=\mathbf{N}
$$

$$
\mathbf{X}=\mathbf{R}^{-1}\mathbf{N}
$$

Let's first define $\mathbf{N}$ as vectors of 1, since we don't want to lose effort by hitting the notoriety ceiling.

In [3]:
def get_X(races):
    R = np.array(relationship_df.loc[races, races])
    invR = np.linalg.inv(R) 
    N = np.ones(len(races))
    X = invR@N
    return X

def get_X_df(races):
    X = get_X(races)
    return pd.DataFrame(list(X)+[sum(X)]+[len(races)/sum(X)], index=races+['sum', 'efficiency']).transpose()

These races selected are the control group of races that are friends, but not necessarily to be the only friends as an notoriety overflow is allowed, which will make $N_i > 1$

To calculate the possible solutions, one has to recuse through all subsets and check if some get possible solutions, see below

## Example solutions

Let's see if there is a solution for all the races

The number for each race represents how much effort you will have to put with that race to get 1 notoriety point for each race assuming you're to be friend with all these selected race.

The sum is the total workload of current tactic.

The efficiency is the sum of actual gained notoriety points (equal the number of races selected) divided by total workload.

In [4]:
get_X_df(all_races)

Teladi      NMMC     Goner  TerraCorp  Strong Arms     Argon     Boron  \
0  0.413665  0.834924  0.400697  -0.602183    -0.088362 -0.559179 -0.534538   

     Split    Atreus      OTAS    Duke's   Paranid   Pirates    Terran  \
0 -0.60492 -0.931595 -1.156118 -0.343422 -1.250485 -1.373975 -1.513874   

       Yaki       sum  efficiency  
0 -1.536987 -8.846352   -1.695614

Unfortunately there is not a solution as one would have to get negative points from races. The problem is, when actively losing points of a race, the notoriety points of enemy races does not increase accordingly. However if that is true, we can see a path that you do some missions for some races but keep robbing other races and one day you will find yourself master of diplomacy with all races - too good to be true for those races you robbed.

Let's get results of some interesting races combo.

In [5]:
commonwealth_races = ["Teladi", "Argon", "Boron", "Split", "Paranid"]
major_races = ["Teladi", "Argon", "Boron", "Split", "Paranid", "Terran"]
all_but_yaki_races = ["Teladi", "NMMC", "Goner", "TerraCorp", "Strong Arms", "Argon", "Boron", "Split", "Atreus", "OTAS", "Duke's", "Paranid", "Pirates", "Terran"]
all_but_yaki_terran_races = ["Teladi", "NMMC", "Goner", "TerraCorp", "Strong Arms", "Argon", "Boron", "Split", "Atreus", "OTAS", "Duke's", "Paranid", "Pirates"]
all_but_yaki_terran_pirates_races = ["Teladi", "NMMC", "Goner", "TerraCorp", "Strong Arms", "Argon", "Boron", "Split", "Atreus", "OTAS", "Duke's", "Paranid"]
all_but_yaki_pirates_races = ["Teladi", "NMMC", "Goner", "TerraCorp", "Strong Arms", "Argon", "Boron", "Split", "Atreus", "OTAS", "Duke's", "Paranid", "Terran"]
all_but_terran_pirates_races = ["Teladi", "NMMC", "Goner", "TerraCorp", "Strong Arms", "Argon", "Boron", "Split", "Atreus", "OTAS", "Duke's", "Paranid", "Yaki"]
all_but_teladi_yaki_terran_pirates_races = ["NMMC", "Goner", "TerraCorp", "Strong Arms", "Argon", "Boron", "Split", "Atreus", "OTAS", "Duke's", "Paranid"]

All commonwealth races, no problem

In [6]:
get_X_df(commonwealth_races)

Teladi     Argon     Boron     Split   Paranid       sum  efficiency
0     1.0  1.428571  1.428571  1.428571  1.428571  6.714286    0.744681

All major races, no problem

In [7]:
get_X_df(major_races)

Teladi     Argon     Boron     Split   Paranid    Terran        sum  \
0     1.0  3.823529  3.823529  3.823529  3.823529  5.588235  21.882353   

   efficiency  
0    0.274194

All but Yaki, not an option

In [8]:
get_X_df(all_but_yaki_races)

Teladi      NMMC    Goner  TerraCorp  Strong Arms     Argon     Boron  \
0  0.904032  0.639788  0.11671  -0.844907    -0.910454 -0.926158 -1.022721   

      Split   Atreus      OTAS    Duke's   Paranid  Pirates    Terran  \
0 -0.939952 -1.63911 -1.890735 -0.748691 -1.995335 -2.18678 -2.677981   

         sum  efficiency  
0 -14.122294    -0.99134

According to calculation just 3 enemies seems not sufficient

In [9]:
get_X_df(all_but_yaki_terran_pirates_races)

Teladi      NMMC     Goner  TerraCorp  Strong Arms      Argon      Boron  \
0 -0.172497  7.816643  7.790471   13.40212    12.541359  10.368289  14.311163   

       Split     Atreus       OTAS     Duke's    Paranid         sum  \
0  17.984757  15.620415  18.506517  22.635896  21.638802  162.443937   

   efficiency  
0    0.073872

which is actually not the case since a subset is sufficient, just teledi notoriety points got overflowed doesn't render it illegal

In [10]:
get_X_df(all_but_teladi_yaki_terran_pirates_races)

NMMC     Goner  TerraCorp  Strong Arms      Argon      Boron  \
0  7.780909  7.781132  13.390187    12.529926  10.355119  14.294295   

       Split     Atreus       OTAS    Duke's    Paranid         sum  \
0  17.964491  15.606104  18.489653  22.60303  21.615287  162.410133   

   efficiency  
0     0.06773

In [11]:
R = np.array(relationship_df.loc[all_races, all_but_teladi_yaki_terran_pirates_races])
R@get_X(all_but_teladi_yaki_terran_pirates_races)

array([  1.16713635,   1.        ,   1.        ,   1.        ,
         1.        ,   1.        ,   1.        ,   1.        ,
         1.        ,   1.        ,   1.        ,   1.        ,
       -24.73473903, -40.03737145, -40.29544981])

thus our complete solution

In [12]:
# https://stackoverflow.com/questions/26332412/python-recursive-function-to-display-all-subsets-of-given-set
def subs(l):
    if l == []:
        return [[]]

    x = subs(l[1:])

    return x + [[l[0]] + y for y in x]

def get_X_optimal_sub(races):
    possible_solution_racess = []
    possible_solution_Xs = []
    possible_solution_workloads = []
    for sub_races in subs(races):
        X = get_X(sub_races)
        R_all = np.array(relationship_df.loc[races, sub_races])
        N_all = R_all@X
        min_N_all = min(N_all)
        if min_N_all>0:
            if min_N_all<1:
                X = [x*1/min_N_all for x in X]
            possible_solution_racess += [sub_races]
            possible_solution_Xs += [X]
            possible_solution_workloads += [sum(X)]
    if len(possible_solution_workloads)>0:
        min_index = possible_solution_workloads.index(min(possible_solution_workloads))
        optiaml_solution_races = possible_solution_racess[min_index]
        optiaml_solution_X = possible_solution_Xs[min_index]
        if min(optiaml_solution_X)>0:
            return pd.DataFrame(list(optiaml_solution_X)+[sum(optiaml_solution_X)]+[len(races)/sum(optiaml_solution_X)], index=optiaml_solution_races+['sum', 'efficiency']).transpose()

In [13]:
get_X_optimal_sub(all_but_yaki_terran_pirates_races)

NMMC     Goner  TerraCorp  Strong Arms      Argon      Boron  \
0  7.780909  7.781132  13.390187    12.529926  10.355119  14.294295   

       Split     Atreus       OTAS    Duke's    Paranid         sum  \
0  17.964491  15.606104  18.489653  22.60303  21.615287  162.410133   

   efficiency  
0    0.073887

In [14]:
get_X_optimal_sub(all_but_yaki_races)

## Let's explore all possibilities

Turned out there are the following number of possible combos.

In [15]:
least_enemy_set_list = []
for sub_races in subs(all_races):
    if sub_races != []:
        X = get_X(sub_races)
        if min(X)>0:
            R_all = np.array(relationship_df.loc[all_races, sub_races])
            N_all = R_all@X
            enemy_set = set([race for i, race in enumerate(all_races) if N_all[i]<=0])
            if enemy_set == set():
                continue
            is_duplicate = False
            for i in reversed(range(len(least_enemy_set_list))):
                least_enemy_set = least_enemy_set_list[i]
                if least_enemy_set.issubset(enemy_set):
                    is_duplicate = True
                    break
                if enemy_set.issubset(least_enemy_set):
                    least_enemy_set_list.pop(i)
            if not is_duplicate:
                least_enemy_set_list = least_enemy_set_list + [enemy_set]
                
len(least_enemy_set_list)

283

which is the following

In [16]:
parsed_least_enemy_set_list = []
for least_enemy_set in least_enemy_set_list:
    workload = sum(get_X(list(set(all_races) - least_enemy_set)))
    efficiency = (len(all_races) - len(list(least_enemy_set)))/workload
    item = " - ".join(list(least_enemy_set)) + ", " + str(workload) + ", " + str(efficiency)
    print(item)
    parsed_least_enemy_set_list += [item]

TerraCorp - Paranid - Teladi - Pirates - Argon - Boron - Split - Strong Arms - NMMC, 1758.3749999999768, 0.0034122414160802327
OTAS - Terran - Atreus - TerraCorp - Teladi - Pirates - Split - Strong Arms - NMMC, 544.4800762025703, 0.011019686967880415
OTAS - Atreus - TerraCorp - Paranid - Teladi - Pirates - Argon - Split - NMMC, 552.8193767538007, 0.010853454586256504
OTAS - Terran - Atreus - TerraCorp - Teladi - Argon - Split - NMMC, 506.48866444611247, 0.01382064494504548
OTAS - Terran - Atreus - TerraCorp - Teladi - Pirates - Argon - NMMC, 15209.636542237093, 0.0004602345348990445
OTAS - Terran - Atreus - TerraCorp - Teladi - Pirates - Boron - NMMC, 618.5248369292519, 0.011317249659290034
OTAS - Terran - Atreus - TerraCorp - Paranid - Teladi - Pirates - NMMC, 11356.756358875922, 0.0006163731772346353
OTAS - Atreus - Teladi - Pirates - Argon - Boron - Split - Strong Arms - NMMC, 479.0827338129511, 0.01252393287532376
OTAS - Terran - Paranid - Teladi - Pirates - Argon - Strong Arms - N

Terran - Atreus - TerraCorp - Pirates - Duke's, 93.52379886462636, 0.10692465577103834
OTAS - Terran - TerraCorp - Pirates - Duke's, 114.8255984182297, 0.08708859468406133
Terran - TerraCorp - Paranid - Pirates - Duke's, 304.55297529784156, 0.03283501003469222
OTAS - Pirates - Argon - Duke's - Boron - Split - Strong Arms, 108.11800528520371, 0.07399322600243001
OTAS - Paranid - Pirates - Argon - Duke's - Boron - Strong Arms, 181.91808462473574, 0.043975836797658464
OTAS - Atreus - Paranid - Argon - Duke's - Split - Strong Arms, 3700.4479519316005, 0.002161900425007754
OTAS - Atreus - Pirates - Argon - Duke's - Split - Strong Arms, 107.5077314220562, 0.07441325283475125
Terran - Atreus - Paranid - Argon - Duke's - Split - Strong Arms, 372.79132624939473, 0.021459726760509598
OTAS - Paranid - Pirates - Argon - Split - Strong Arms - NMMC, 1808.0046528839068, 0.004424767373933127
OTAS - Atreus - Paranid - Pirates - Argon - Duke's - Strong Arms, 179.49442107032894, 0.044569630366759226
OTAS

3 or less enemies option are listed below

In [17]:
parsed_least_enemy_set_list = []
for least_enemy_set in least_enemy_set_list:
    if len(list(least_enemy_set)) <= 3:
        workload = sum(get_X(list(set(all_races) - least_enemy_set)))
        efficiency = (len(all_races) - len(list(least_enemy_set)))/workload
        item = " - ".join(list(least_enemy_set)) + ", " + str(workload) + ", " + str(efficiency)
        print(item)
        parsed_least_enemy_set_list += [item]

Terran - Yaki - Paranid, 195.74987462921916, 0.06130272125450846
Pirates - Terran - Yaki, 162.4439368085863, 0.07387163987622415


4 enemies

In [18]:
parsed_least_enemy_set_list = []
for least_enemy_set in least_enemy_set_list:
    if len(list(least_enemy_set)) == 4:
        workload = sum(get_X(list(set(all_races) - least_enemy_set)))
        efficiency = (len(all_races) - len(list(least_enemy_set)))/workload
        item = " - ".join(list(least_enemy_set)) + ", " + str(workload) + ", " + str(efficiency)
        print(item)
        parsed_least_enemy_set_list += [item]

Terran - Argon - TerraCorp - Yaki, 6380.360232707097, 0.0017240405868639889
Terran - Boron - TerraCorp - Yaki, 1167.9560442782522, 0.009418162655939281
Terran - Atreus - TerraCorp - Yaki, 174.70372019814636, 0.0629637422003605
OTAS - Yaki - TerraCorp - Terran, 178.8162569639823, 0.06151565963163882
Terran - Duke's - TerraCorp - Yaki, 1169.8407341546074, 0.009402989380387082
Strong Arms - OTAS - Terran - Yaki, 281.97836679426274, 0.039010084798547076
OTAS - Argon - Terran - Yaki, 83.24190571176662, 0.13214498041513603
Terran - Atreus - Boron - Yaki, 362.032699777006, 0.030383995718550976
OTAS - Yaki - Boron - Terran, 78.34237671512541, 0.14040932201992093
OTAS - Yaki - Terran - Split, 121.09694482534427, 0.09083631313626518
OTAS - Atreus - Terran - Yaki, 151.90811995639712, 0.07241219233808818
Terran - Duke's - Atreus - Yaki, 134248.85468452683, 8.193738431399691e-05
OTAS - Duke's - Terran - Yaki, 68.01331847910072, 0.16173302885346058
OTAS - NMMC - Terran - Yaki, 824.8106996204865, 0.0

4 enemies but without terran

In [19]:
parsed_least_enemy_set_list = []
for least_enemy_set in least_enemy_set_list:
    if len(list(least_enemy_set)) == 4 and "Terran" not in least_enemy_set:
        workload = sum(get_X(list(set(all_races) - least_enemy_set)))
        efficiency = (len(all_races) - len(list(least_enemy_set)))/workload
        item = " - ".join(list(least_enemy_set)) + ", " + str(workload) + ", " + str(efficiency)
        print(item)
        parsed_least_enemy_set_list += [item]

5 enemies but without any major races

In [20]:
parsed_least_enemy_set_list = []
for least_enemy_set in least_enemy_set_list:
    if len(list(least_enemy_set)) == 5 and set(major_races).intersection(least_enemy_set) == set():
        workload = sum(get_X(list(set(all_races) - least_enemy_set)))
        efficiency = (len(all_races) - len(list(least_enemy_set)))/workload
        item = " - ".join(list(least_enemy_set)) + ", " + str(workload) + ", " + str(efficiency)
        print(item)
        parsed_least_enemy_set_list += [item]

Atreus - Pirates - Duke's - Strong Arms - Yaki, 115.80946764695183, 0.08634872608589532
OTAS - Pirates - Duke's - Strong Arms - Yaki, 55.18972548571193, 0.18119314622409022
Atreus - TerraCorp - Pirates - Duke's - Yaki, 280.2498134147285, 0.03568245016171151
OTAS - Atreus - Pirates - Duke's - Yaki, 59.131288178803246, 0.16911520631449212


In [21]:
parsed_least_enemy_set_list = []
for least_enemy_set in least_enemy_set_list:
    if "Yaki" in least_enemy_set:
        workload = sum(get_X(list(set(all_races) - least_enemy_set)))
        efficiency = (len(all_races) - len(list(least_enemy_set)))/workload
        item = " - ".join(list(least_enemy_set)) + ", " + str(workload) + ", " + str(efficiency)
        print(item)
        parsed_least_enemy_set_list += [item]
        
len(parsed_least_enemy_set_list)

TerraCorp - Pirates - Argon - Duke's - Strong Arms - Yaki, 21962.27452879679, 0.00040979362079274895
TerraCorp - Pirates - Duke's - Boron - Strong Arms - Yaki, -27986.687748995337, -0.0003215814633270806
Pirates - Duke's - Split - Strong Arms - Yaki, 157.46622151477678, 0.06350568333832529
Atreus - Pirates - Duke's - Strong Arms - Yaki, 115.80946764695183, 0.08634872608589532
OTAS - Pirates - Duke's - Strong Arms - Yaki, 55.18972548571193, 0.18119314622409022
Paranid - Pirates - Duke's - Strong Arms - Yaki, 76.97766506101125, 0.12990781146809482
Atreus - TerraCorp - Argon - Boron - Split - Strong Arms - Yaki, 173.9153947095792, 0.045999378107724025
TerraCorp - Pirates - Argon - Boron - Split - Strong Arms - Yaki, 104.81626022629128, 0.07632403581971477
OTAS - Atreus - TerraCorp - Argon - Split - Strong Arms - Yaki, 651.6304134884892, 0.0122768978156992
OTAS - TerraCorp - Argon - Duke's - Split - Strong Arms - Yaki, 98.13427898713863, 0.0815209535604625
TerraCorp - Paranid - Argon - Spl

OTAS - Atreus - Paranid - Split - NMMC - Yaki, 586.2996636003334, 0.015350511963000354
OTAS - NMMC - Terran - Yaki, 824.8106996204865, 0.013336393435562052


116

## Conclusion

- A simple tool is developed
- All 283 possible combos are given with total workload and efficiency
- The only two options to maintain at most 3 enemies (races, workload, efficiency)
    - Terran - Yaki - Paranid, 195.74987462921916, 0.06130272125450846
    - Pirates - Terran - Yaki, 162.4439368085863, 0.07387163987622415
- 4 options to keep good relationship with all 6 major races with least enemies (5) are listed as below:
    - Atreus - Pirates - Duke's - Strong Arms - Yaki, 115.80946764695183, 0.08634872608589532
    - OTAS - Pirates - Duke's - Strong Arms - Yaki, 55.18972548571193, 0.18119314622409022
    - Atreus - TerraCorp - Pirates - Duke's - Yaki, 280.2498134147285, 0.03568245016171151
    - OTAS - Atreus - Pirates - Duke's - Yaki, 59.131288178803246, 0.16911520631449212